### DISCUSSION TITLE
Custom Test Issue and Understanding Grandparent Rollups

### DISCUSSION THREAD

**Discussion by rygramer** [jamessimone/apex-rollup#178] [Q&A]
Hey @jamessimone, sorry to bother you again. Moving this over to a discussion rather than an issue because it's entirely possible I'm missing the boat somewhere.

I'm having trouble understanding how to configure Grandparent Rollups: https://share.vidyard.com/watch/UudmSyaRa27C7dDQPWWC5z?

And I mentioned that I wanted to write some tests to better understand, but got hung up on the most basic test:
```
@isTest
public class RollupTest {
    @isTest
    public static void test1(){
    	Grandparent__c gp = new Grandparent__c(
            AmountfromChildren__c = 0
        );
        insert gp;
        
        Parent__c p = new Parent__c(
            Grandparent__c = gp.Id,
            AmountfromChildren__c = 0
        );
        insert p;
        
        Child__c c = new Child__c(
            Parent__c = p.Id
        );
        insert c;
        
        test.startTest();
        
        c.Amount__c = 5;
        update c;
        
        test.stopTest();
        
       // I would expect this to return 5.00, but it returns 0.00
        system.debug([SELECT AmountfromChildren__c FROM Parent__c LIMIT 1]);
    }
}
```
Log from that test - 
[apex-07L780000008vQ2EAI.log](https://github.com/jamessimone/apex-rollup/files/7245180/apex-07L780000008vQ2EAI.log)

**Comment by rygramer** [comment]
FWIW, I moved the logic from flow to triggers and it works perfectly. 

I did find a slight error in my configuration - in Grandparent Relationship Field Path, I put `Parent__r.Grandparent__r.AmountfromChildren__c`, but it should have been `Parent__r.GrandParent__r.AmountfromChildren__c` (capital P in Grand**P**arent). Capitalization is important.

The test still isn't working as expected.

**Reply by jamessimone** [reply]
@rygramer this is a really great test, and it’s my hope that writing unit tests - in conjunction with the Apex Logs the running of an individual test produces - will help to show you what the issue is (at least, what I suspect the issue is given the test you’ve just shown me).

if you were to check off `Is Rollup Logging Enabled` on your Org Default Rollup Control CMDT record, and open up the Apex Log generated (by default, if you have the Salesforce Dev Console open when running tests; you can also configure VS Code to start a trace for your user) when running that test, I believe you’ll see a message declaring this rollup to be a no-op, and an early exit.

If you compare the test you’ve written to the tests in any of the Rollup test classes (pay particular attention to the test setup methods), I’m willing to bet there’s a _very_ small change you can make that will get your test passing. Do you see it?

**Reply by jamessimone** [reply]
Ah, also, just seeing the log you pasted (apologies; up to speed now). The answer above still applies though 😁

**Reply by rygramer** [reply]
Ah, yes, but of course 😅

`upsert new RollupSettings__c(IsEnabled__c = true);`

**Reply by jamessimone** [reply]
That’s the one!! There you go

**Reply by clint-rosetree** [reply]
I had this same issue and fought it for hours. Is this documented somewhere that I missed?

**Reply by jamessimone** [reply]
@clint-rosetree it's one of the first things [mentioned in the README](https://github.com/jamessimone/apex-rollup#deployment--setup). For people writing unit tests against Apex Rollup, how do you think I can best document this feature? Should I have another section in the README?

**Reply by clint-rosetree** [reply]
Yeah, I followed the README for installation and everything worked live, but it just didn't cross my mind while later re-writing legacy test classes that the custom setting wouldn't be accessible. I've grown too accustomed to CMDT, I guess.

I would have definitely benefited from a simple section about writing test classes against the framework that just mentioned we have to include that one line in the @testSetup.

Thanks James!

**Reply by jamessimone** [reply]
Excellent feedback - I'll include that callout within the README in the next version. Thanks!

**Reply by jamessimone** [reply]
[Here's the updated README](https://github.com/jamessimone/apex-rollup/commit/ddc3122d881c91363d013244e1179ea036183c7e) - let me know if either of you have feedback. Thanks again for the nudge, I always appreciate guidance on improving the documentation (which I know is quite dense, already! 😅)
